### PR TITLE
Refactor build system, simplify templates and escape macros in code

### LINKED ISSUE #50: @@-substitutions break validity of plugins' sources
Plugin's sources now are not valid js code, because of build-time substitution blocks like `@@METAINFO@@`, `@@PLUGINSTART@@`, `@@PLUGINEND@@`.

It complicates applying such tools like `eslint`.

It would be great if we could rearrange the code to make the substitutions appear commented, e.g.
```
// @@METAINFO@@
// ...
// @@PLUGINSTART@@
// ...
// @@PLUGINEND@@
```

### LINKED ISSUE #99: Scripts versioning
Currently every version number contains date/time stamp as it's last component.

So with every build every script get newer version, even if it's content stays the same.

With this system if we turn on auto-updating feature - it will fire unconditionally, on every new build.

I don't think that it is good design, as typically we want to update changed scripts only.

### PR DESCRIPTION
**Pull request by johnd0e**
- Refactor build system into several modules (can be used as cli utilities). Implement `watch` mode to auto-rebuild on source changes.
- Escape macros in code in order to keep js-validity, fix #50.
  Use template from settings instead of url harcoding (closes #150).
- Simplify userscripts source template, rename plugin sources:  `*.user.js` -> `*.js`, rearrange source directories.
- Remove timestamp component from `version` of Release scripts (closes #99).
- Additional dev tools: `web_meta_gen.py`, `web_server_local.py`, `tampermonkey_stubs_gen.py`.

### CODE REVIEW

**Comment by johnd0e** [comment]
.

File: build.py [hunk 1]
```diff
@@ -20,19 +20,18 @@
 # load settings file
 from buildsettings import buildSettings
 
+defaultBuild = None
 # load option local settings file
-try:
+if os.path.isfile('./localbuildsettings.py'):
```
Comment (modos189) [inline build.py]: If file is missing, there will be an exception. Not great, not terrible.

BTW, we can use:
```
# load option local settings file and default build
try:
    from localbuildsettings import buildSettings as localBuildSettings, defaultBuild
    buildSettings.update(localBuildSettings)
except ImportError:
    localBuildSettings = {}
    defaultBuild = None
```

**Comment by johnd0e** [comment]
I am also going to enhance build script to be able to build single plugin (even 3rd-party).

File: code/boot.js [hunk 2]
```diff
@@ -829,23 +829,23 @@ window.registerMarkerForOMS = function(marker) {
 }
 
 try {
-@@INCLUDERAW:external/autolink-min.js@@
+@include_raw:external/autolink-min.js@
```
Comment (modos189) [inline code/boot.js]: Maybe we should add quotes, make a string?

File: plugins/debug-raw-portal-data.user.js [hunk 3]
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-raw-portal-data
```
Comment (modos189) [inline plugins/debug-raw-portal-data.user.js]: Did you leave your id here on purpose, or did you forget to remove it?

File: plugins/periodic-refresh.user.js [hunk 4]
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-periodic-refresh
```
Comment (modos189) [inline plugins/periodic-refresh.user.js]: This is probably where you forgot to remove ID

File: total-conversion-build.js [hunk 5]
```diff
@@ -1,13 +1,10 @@
...
-// @id             ingress-intel-total-conversion@jonatkins
+// @author         jonatkins
 // @name           IITC: Ingress intel map total conversion
-// @version        0.29.1.@@DATETIMEVERSION@@
-// @description    [@@BUILDNAME@@-@@BUILDDATE@@] Total conversion for the ingress intel map.
-@@METAINFO@@
+// @version        0.29.1
+// @description    Total conversion for the ingress intel map.
 // @run-at         document-end
-// ==/UserScript==
```
Comment (modos189) [inline total-conversion-build.js]: BTW, maybe we should move this file to `code` folder?

File: plugins/fix-china-map-offset.user.js [hunk 6]
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-fix-china-map-offset
+﻿// @id             iitc-plugin-fix-china-map-offset
```
Comment (modos189) [inline plugins/fix-china-map-offset.user.js]: again

File: build_plugin.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    meta, is_plugin = fill_meta(meta, name, dist_path)
+    settings.plugin_id = name
+
+    filename = os.path.join(out_dir, name) if out_dir else name
+    if settings.update_file == '.meta.js':
+        io.open(filename + settings.update_file, 'w', encoding='utf8').write(meta)
+
+    script = re.sub(r"'@bundle_code@';", bundle_code, script)
+    wrapper = import_module(settings.plugin_wrapper)
+    template = r"'@(\w+)(?::([0-9a-zA-Z_./-]+))?@'"
```
Comment (modos189) [inline build_plugin.py]: Ooh. Uh...)
At least add a comment, please, that it's for finding `@type:string@` tags

File: build_plugin.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        if settings.update_file in {'.user.js', '.meta.js'}:
+            append_line('updateURL', path + settings.update_file)
+        append_line('downloadURL', path + '.user.js')
+    append_line('id', name)
+    append_line('namespace', settings.namespace)
+    if keys.isdisjoint({'match', 'include'}):
+        append_line('match', settings.match)
+    append_line('grant', 'none')
+    meta.append('// ==/UserScript==\n')
+    return '\n'.join(meta), is_plugin
```
Comment (modos189) [inline build_plugin.py]: Please add blank lines, everything is now merging into a solid block

File: plugins/basemap-gaode.user.js [hunk 9]
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-basemap-gaode
+﻿// @id             iitc-plugin-basemap-gaode
```
Comment (modos189) [inline plugins/basemap-gaode.user.js]: id?

File: plugins/basemap-kartverket.user.js [hunk 10]
```diff
@@ -1,15 +1,8 @@
-﻿// ==UserScript==
-// @id             iitc-plugin-basemap-kartverket
+﻿// @id             iitc-plugin-basemap-kartverket
```
Comment (modos189) [inline plugins/basemap-kartverket.user.js]: id?

File: plugins/link-show-direction.user.js [hunk 11]
```diff
@@ -1,15 +1,8 @@
-// ==UserScript==
 // @id             iitc-plugin-link-show-direction
```
Comment (modos189) [inline plugins/link-show-direction.user.js]: id?

**Review by modos189 — DISMISSED** [review]
It's pretty cool on the whole. I left a few questions in the comments

File: plugins/bookmarks.css [hunk 12]
```diff
@@ -280,7 +280,7 @@
 	cursor:pointer;
 }
 #bkmrksTrigger, .bkmrksStar span, .portal-list-bookmark span {
-	background-image:url(@@INCLUDEIMAGE:images/bookmarks-img.png@@);
+	background-image:url(images/bookmarks-img.png);
```
Comment (modos189) [inline plugins/bookmarks.css:283]: Should we use `@include_img` here?

**Review by modos189 — CHANGES_REQUESTED** [review]
Good job!

File: plugins/bookmarks.css (hunk 12 above)
Comment (johnd0e) [reply plugins/bookmarks.css:283]: No, that is handled in another way.
Urls in CSS are easy to extract with regexp, so no need for special templates.  

See corresponding change in `bookmarks.js`.

Comment (modos189) [reply plugins/bookmarks.css:283]: Ok

**Review by modos189 — APPROVED** [review]

**Pull request merged**